File: Refinement/Modules/AnalyzeBids.ts
import { BidAnalysisData, CleanBiddingData, ScrapedData } from "../../Contracts/ExtractionContracts.js";
import { binomial, determineCDF, determineProbabilityMassDistribution, getProbabilityFromCDF } from "../../Utilities/Stats.js";

export const analyzeBidDistributions = (data: ScrapedData): BidAnalysisData => {
    const gameData = Object.values(data)[0]
    if (!gameData) throw "Empty json?"

    const bids: CleanBiddingData[] = gameData.TrackBids

    const ironThroneData = bids.filter((bidData) => bidData.Track == "Iron Throne");
    const fiefdomData = bids.filter((bidData) => bidData.Track == "Fiefdom");
    const kingsCourtData = bids.filter((bidData) => bidData.Track == "King's Court");

    const bidDataAccessor = (element : CleanBiddingData) => element.Amount;

    const ironThroneDistribution = determineProbabilityMassDistribution(ironThroneData, bidDataAccessor);
    const fiefdomDistribution = determineProbabilityMassDistribution(fiefdomData, bidDataAccessor);
    const kingsCourtDistribution = determineProbabilityMassDistribution(kingsCourtData, bidDataAccessor);

    const averageBid =
        bids.length > 0
            ? bids.reduce((sum, bid) => sum + bid.Amount, 0) / bids.length
            : 0;

    return {
        "Iron Throne Distribution": ironThroneDistribution,
        "Fiefdom Distribution": fiefdomDistribution,
        "King's Court Distribution": kingsCourtDistribution,
        "Iron Throne Bid Chart": determineBiddingPlacement(ironThroneDistribution),
        "Fiefdom Bid Chart": determineBiddingPlacement(fiefdomDistribution),
        "King's Court Bid Chart": determineBiddingPlacement(kingsCourtDistribution),
        "Average Bid": averageBid,
    };
};

// Returns one series per placement (index 0 = 1st place, 1 = 2nd place, …).
// Each series contains { amount, probability } for every bid amount, letting
// you read off: "to finish Nth, bid X gives you Y% success".
const determineBiddingPlacement = (BidPartialMassDistribution: {[key: number]: number}, playerCount: number = 8, maxBid = 20) => {
    const CDF = determineCDF(BidPartialMassDistribution);
    const numOpponents = playerCount - 1;

    const placementCharts: { amount: number; probability: number }[][] =
        Array.from({ length: playerCount }, () => []);

    for (let bid = 0; bid <= maxBid; bid++) {
        const probability = getProbabilityFromCDF(bid, CDF);

        for (let outbidBy = 0; outbidBy <= numOpponents; outbidBy++) {
            // outbidBy = number of opponents who bid strictly higher than you
            const probOfRank =
                binomial(numOpponents, outbidBy)
                * Math.pow(1 - probability, outbidBy)
                * Math.pow(probability, numOpponents - outbidBy);

            placementCharts[outbidBy]!.push({ amount: bid, probability: probOfRank });
        }
    }

    return placementCharts;
}

File: 1_Analysis/Modules/AnalyzeBids.py
from typing import Dict, Any, List, cast

from Contracts.ExtractionContracts import ScrapedData
from Utilities.Stats import (
    determine_probability_mass_distribution,
    determine_cdf,
    get_probability_from_cdf,
    binomial,
)


def analyze_bid_distributions(data: ScrapedData) -> Dict[str, Any]:
    inner: Dict[str, Any] = list(data.values())[0]
    if not inner:
        raise ValueError("Empty json?")

    bids: List[Dict[str, Any]] = cast(List[Dict[str, Any]], inner['TrackBids'])

    iron_throne_data = [b for b in bids if b['Track'] == 'Iron Throne']
    fiefdom_data = [b for b in bids if b['Track'] == 'Fiefdom']
    kings_court_data = [b for b in bids if b['Track'] == "King's Court"]

    def bid_accessor(element: Dict[str, Any]) -> float:
        return float(element['Amount'])

    iron_throne_dist = determine_probability_mass_distribution(iron_throne_data, bid_accessor)
    fiefdom_dist = determine_probability_mass_distribution(fiefdom_data, bid_accessor)
    kings_court_dist = determine_probability_mass_distribution(kings_court_data, bid_accessor)

    avg_bid = sum(float(b['Amount']) for b in bids) / len(bids) if bids else 0.0

    return {
        'Iron Throne Distribution': iron_throne_dist,
        'Fiefdom Distribution': fiefdom_dist,
        "King's Court Distribution": kings_court_dist,
        'Iron Throne Bid Chart': _determine_bidding_placement(iron_throne_dist),
        'Fiefdom Bid Chart': _determine_bidding_placement(fiefdom_dist),
        "King's Court Bid Chart": _determine_bidding_placement(kings_court_dist),
        'Average Bid': avg_bid,
    }


def _determine_bidding_placement(
    bid_partial_mass_distribution: Dict[float, float],
    player_count: int = 8,
    max_bid: int = 20,
) -> List[List[Dict[str, Any]]]:
    cdf = determine_cdf(bid_partial_mass_distribution)
    num_opponents = player_count - 1

    placement_charts: List[List[Dict[str, Any]]] = [[] for _ in range(player_count)]

    for bid in range(max_bid + 1):
        probability = get_probability_from_cdf(bid, cdf)

        for outbid_by in range(num_opponents + 1):
            prob_of_rank = (
                binomial(num_opponents, outbid_by)
                * (1 - probability) ** outbid_by
                * probability ** (num_opponents - outbid_by)
            )
            placement_charts[outbid_by].append({'amount': bid, 'probability': prob_of_rank})

    return placement_charts
